### PR TITLE
Feature javascript refactor V1 (data.js + css naming/indices)

### PR DESCRIPTION
**Pull request by mcgill-a**

### CODE REVIEW

**Review by athompsonScottLogic — APPROVED** [review]
cleaner

File: whichone/static/js/data.js [hunk 1]
```diff
@@ -1,103 +1,68 @@
...
+
+function getLocalData(user_id) {
+  // get data from local storage instead of the web server
+  let data = localStorage.getItem(user_id);
+  let local = JSON.parse(data);
 
-    if (resetModes) {
-      user.modes = {
+  // check if modes are present and valid
+  if (local.modes != null) {
```
Comment (bdimitrov-SL) [inline whichone/static/js/data.js:18]: use strict equality check

File: whichone/static/js/data.js [hunk 2]
```diff
@@ -1,103 +1,68 @@
...
-        return false;
-      }
-    } else {
-      return false;
-    }
-
-    return true;
-  } else {
-    return false;
+  if (local.high_score == null || Number.isNaN(local.high_score)) {
```
Comment (bdimitrov-SL) [inline whichone/static/js/data.js:32]: use strict equality check

File: whichone/static/js/main.js [hunk 3]
```diff
@@ -37,22 +38,31 @@ var refreshIntervalId = null;
...
-  if (!localDataFound()) {
-    getSpotifyData();
-  }
-
+  initData();
   initOptions();
   setMuteIcon();
 });
 
+async function initData(currentTime = Date.now()) {
```
Comment (bdimitrov-SL) [inline whichone/static/js/main.js:47]: No need to be async

File: whichone/static/js/ui.js [hunk 4]
```diff
@@ -100,7 +100,7 @@ function getGameOverText() {
   }
 }
 
-function wrongAnswerUI(option) {
+async function wrongAnswerUI(option) {
```
Comment (bdimitrov-SL) [inline whichone/static/js/ui.js:103]: No need to be async

File: whichone/static/js/data.js [hunk 5]
```diff
@@ -1,103 +1,68 @@
...
 }
 
-async function getAudioFeatures() {
-  let trackIds = [];
-  if (user.top_tracks != null && user.top_tracks.length > 0) {
-    user.top_tracks.forEach((track) => {
-      trackIds.push(track.id);
-    });
-
+async function post_request(url, track_ids) {
```
Comment (bdimitrov-SL) [inline whichone/static/js/data.js:38]: No need to be async

File: whichone/static/js/data.js [hunk 6]
```diff
@@ -1,103 +1,68 @@
...
-      timeout: 3000,
     });
-  } else {
-    console.error("Audio features unavailable (no tracks loaded).");
-  }
+  });
 }
 
-async function request(key) {
+async function get_request(url) {
```
Comment (bdimitrov-SL) [inline whichone/static/js/data.js:59]: No need to be async

File: whichone/static/js/data.js [hunk 7]
```diff
@@ -107,18 +72,42 @@ async function request(key) {
...
+}
 
-  request("top_tracks")
-    .then(() => {
-      getAudioFeatures();
-    })
-    .catch(() => {
-      console.error("API unavailable. Please try again later.");
-    });
+async function getSpotifyData(
```
Comment (bdimitrov-SL) [inline whichone/static/js/data.js:83]: No need to be async

File: whichone/static/js/data.js (hunk 5 above)
Comment (connor-settle) [reply whichone/static/js/data.js:38]: @mcgill-a @athompsonScottLogic You need async only if you are awaiting within the function. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function

File: whichone/static/js/data.js [hunk 8]
```diff
@@ -107,18 +72,42 @@ async function request(key) {
...
 
-async function getSpotifyData() {
-  request("top_artists").catch(() => {
-    console.error("API unavailable. Please try again later.");
+function processAudioFeatures(audio_features) {
+  let output = {};
+  audio_features.forEach((result) => {
+    output[result["id"]] = result;
   });
+  return output;
```
Comment (connor-settle) [inline whichone/static/js/data.js:80]: Consider [reduce](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/Reduce) to clean this up

File: whichone/static/js/main.js (hunk 3 above)
Comment (connor-settle) [inline whichone/static/js/main.js:47]: Better to inject a factory than a optional param.

You are assuming currentTime will be used immediately but better to not make that assumption :)

File: whichone/static/js/ui.js [hunk 9]
```diff
@@ -164,23 +149,23 @@ function correctAnswerUI(option) {
...
-    }, opacityDelay);
-  }, 1000);
+  await new Promise((resolve) => setTimeout(resolve, primaryDelay));
+  $(".choice").css("opacity", 0);
+  $(".choice").css("cursor", "default");
+  $(".down").css("opacity", 0);
+  $("#stats-popup").css("opacity", 0);
+
+  // after cards have faded, display the stats popup
+  await new Promise((resolve) => setTimeout(resolve, secondaryDelay));
```
Comment (connor-settle) [inline whichone/static/js/ui.js:164]: Marzio's suggestion here might be useful elsewhere. This function looks clean & has no nesting. 

async/await allows us to write code as though it is synchronous.

**Comment by PHardySL** [comment]
issues fixed in feature-js-modules branch

**Pull request closed without merging**